RockLib.Messaging/DependencyInjection/MessagingExtensions.cs

```diff
@@ -45,6 +46,60 @@ public static class MessagingExtensions
+        /// Adds an <see cref="ISender"/> to the service collection where the sender is created
+        /// with options obtained by name from an <see cref="IOptionsMonitor{TOptions}"/> using
+        /// the specified factory method.
+        /// <para>If <paramref name="reloadOnChange"/> is true, then the sender will automatically
+        /// reload itself whenever its options
+        /// changes.</para>
+        /// </summary>
+        /// <typeparam name="TSenderOptions">The type of options used to create the sender.</typeparam>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="senderName">
+        /// The name that identifies which named options to use from the <see cref=
+        /// "IOptionsMonitor{TOptions}"/>.
+        /// </param>
+        /// <param name="createSender">
+        /// A function that creates the instance of <see cref="ISender"/> using options obtained
+        /// from the <see cref="IOptionsMonitor{TOptions}"/>.
+        /// </param>
+        /// <param name="configureOptions">
+        /// An optional callback for configuring the <typeparamref name="TSenderOptions"/>. If
+        /// provided, invoking this callback is the <em>last</em> step in creating and configuring
+        /// the options used to create the sender.
+        /// </param>
+        /// <param name="reloadOnChange">
+        /// Whether to create a sender that automatically reloads itself when its configuration
+        /// or options change.
+        /// </param>
+        /// <param name="lifetime">The <see cref="ServiceLifetime"/> of the sender.</param>
+        /// <returns>A new <see cref="ISenderBuilder"/> for decorating the <see cref="ISender"/>.</returns>
+        public static ISenderBuilder AddSender<TSenderOptions>(this IServiceCollection services, string senderName,
+            Func<TSenderOptions, IServiceProvider, ISender> createSender, Action<TSenderOptions> configureOptions = null,
+            bool reloadOnChange = true, ServiceLifetime lifetime = _defaultLifetime)
+            where TSenderOptions : class, new()
+        {
+            if (services is null)
+                throw new ArgumentNullException(nameof(services));
+            if (senderName is null)
+                throw new ArgumentNullException(nameof(senderName));
+            if (createSender is null)
+                throw new ArgumentNullException(nameof(createSender));
+
+            return services.AddSender(serviceProvider =>
+            {
+                var optionsMonitor = serviceProvider.GetService<IOptionsMonitor<TSenderOptions>>();
+                var options = optionsMonitor?.Get(senderName) ?? new TSenderOptions();
+                configureOptions?.Invoke(options);
+
+                if (reloadOnChange && optionsMonitor != null)
+                    return new ReloadingSender<TSenderOptions>(serviceProvider, senderName, createSender, options, optionsMonitor, configureOptions);
```

Comment:
Passing this `serviceProvider` through the constructor and only using it forward on to `createSender` feels kinda tedious. We have this extra property that were holding but never really using in any meaningful way.

Could `createSender` be reverted back to its single-argument version as far as `ReloadingSender` is concerned, and here we just bake `serviceProvider` in? We could pass `o => createSender(o, serviceProvider)` to this constructor instead.

---

I have similar feelings about passing in the initial `options` value. It feels a bit redundant, and I wonder if the constructor should just pull out the initial value instead.

Reply:
Good points about the superfluous constructor parameters. Updated.